docs/proposals/20210524-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,573 @@
+
+    // +optional
+    Version string `json:"version,omitempty"`
+
+    // +optional
+    EtcdadmArgs map[string]interface{} `json:"etcdadmArgs,omitempty"`
+}
+
+
+type EtcdadmConfigStatus struct {
+    Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+    
+    DataSecretName *string `json:"dataSecretName,omitempty"`
+
+    Ready bool `json:"ready,omitempty"`
+}
+```
+
+The following fields will be added/modified in existing CAPI types:
+- Add Cluster.Spec.ManagedExternalEtcdRef as:
+```go
+// ManagedExternalEtcdRef is an optional reference to an etcd provider resource that holds details
+// for provisioning an external etcd cluster
+// +optional
+ManagedExternalEtcdRef *corev1.ObjectReference `json:"managedExternalEtcdRef,omitempty"`
+```
+- Add Cluster.Status.ManagedExternalEtcdReady field as:
+```go
+// ManagedExternalEtcdReady indicates external etcd cluster is fully provisioned
+// +optional
+ManagedExternalEtcdReady bool `json:"managedExternalEtcdReady"`
+```
+- Make KubeadmConfigSpec.ClusterConfiguration.Etcd.External.Endpoints field mutable
+    - The kubeadm validating webhook currently doesn't allow this field to get updated. It should be added as an [allowed path](https://github.com/mrajashree/cluster-api/commit/18c42c47d024ce17cdda39500fc0d6bd67c5aa67#diff-6603d336435f3ee62a50043413f62441b685d5c416e5cadd792ed2a536b2f55fR121) in the webhook.
+
+
+#### Implementation details
+
+##### Contract for pluggable etcd providers
+This section defines a contract for an etcd provider. Any etcd provider being added should follow this contract. We are starting by adding an etcdadm based provider, but it is possible that later someone wants to add a new provider that uses something else to bring up etcd. Hence we will design it in a way that's pluggable.
+
+###### An etcd controller’s main responsibilities are:
+
+* Managing a set of Machines that represent an etcd member.
+* Managing the etcd cluster, in terms of scaling up/down/upgrading etcd version, and running healthchecks.
+* Populating and updating a field that will contain the etcd cluster endpoints (IP addresses of all etcd members). 
+* Creating/managing two Kubernetes Secrets that will contain certs required by the API server to communicate with the etcd cluster
+    * One Secret containing the etcd CA cert, by the name `{cluster.Name}-etcd`
+    * One Secret containing the API server cert-key pair, by the name `{cluster.Name}-apiserver-etcd-client`
+
+###### Required services:
+The etcd provider should install etcd with the version specified by the user for each member.
+
+###### Required fields:
+Each etcd provider should add a new API type to manage the etcd cluster lifecycle, and it should have the following required fields:
+- Spec:
+    - Replicas: Allow only odd numbers, and immutable after creation.
+- Status
+    - Ready: To be set to true once all etcd members pass healthcheck (making calls to /health endpoint).
```

Comment:
This is `Status` that represents the whole cluster? What is the action when this evaluates to `false`? Or just read-only mode for visibility?

Reply:
makes sense to not name it "Ready", changed it to "CreationComplete"